src/infra/core/src/repos/metadata_chain_validators.rs

```diff
@@ -323,3 +339,553 @@ impl<'a> MetadataChainVisitor for ValidateOffsetsAreSequentialVisitor<'a> {
+        if let MetadataEvent::SetTransform(e) = &block.event {
+            // Ensure has inputs
+            if e.inputs.is_empty() {
+                invalid_event!(e.clone(), "Transform must have at least one input");
+            }
+
+            // Ensure inputs are resolved to IDs and aliases are specified
+            for i in &e.inputs {
+                if i.dataset_ref.id().is_none() || i.alias.is_none() {
+                    invalid_event!(
+                        e.clone(),
+                        "Transform inputs must be resolved to dataset IDs and specify aliases"
+                    );
+                }
+            }
+
+            // Queries must be normalized
+            validate_transform(&block.event, &e.transform)?;
+        }
+
+        Ok(Self {})
+    }
+}
+
+impl MetadataChainVisitor for ValidateSetTransformVisitor {
+    type Error = AppendError;
+
+    fn initial_decision(&self) -> Decision {
+        Decision::Stop
+    }
+
+    fn visit(&mut self, _: HashedMetadataBlockRef) -> Result<Decision, Self::Error> {
+        unreachable!()
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct ValidateEventIsNotEmptyVisitor {}
+
+impl ValidateEventIsNotEmptyVisitor {
+    pub fn new(block: &MetadataBlock) -> Result<Self, AppendError> {
+        match &block.event {
+            // TODO: ensure only used on Root datasets
+            MetadataEvent::AddData(e) if e.is_empty() => {
+                return Err(AppendValidationError::empty_event(e.clone()).into())
+            }
+            // TODO: ensure only used on Derivative datasets
+            MetadataEvent::ExecuteTransform(e) if e.is_empty() => {
+                return Err(AppendValidationError::empty_event(e.clone()).into())
+            }
+            _ => (),
+        };
+
+        Ok(Self {})
+    }
+}
+
+impl MetadataChainVisitor for ValidateEventIsNotEmptyVisitor {
+    type Error = AppendError;
+
+    fn initial_decision(&self) -> Decision {
+        Decision::Stop
+    }
+
+    fn visit(&mut self, _: HashedMetadataBlockRef) -> Result<Decision, Self::Error> {
+        unreachable!()
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct ValidateAddDataVisitor<'a> {
+    appended_add_data: Option<&'a AddData>,
+    prev_schema: Option<SetDataSchema>,
+    prev_add_data: Option<AddData>,
+    next_block_flags: Flag,
+}
+
+impl<'a> ValidateAddDataVisitor<'a> {
+    pub fn new(block: &'a MetadataBlock) -> Result<Self, AppendError> {
+        let appended_add_data = match &block.event {
+            MetadataEvent::AddData(e) => Some(e),
+            _ => None,
+        };
+
+        Ok(Self {
+            appended_add_data,
+            prev_schema: None,
+            prev_add_data: None,
+            next_block_flags: Flag::SET_DATA_SCHEMA | Flag::ADD_DATA,
+        })
+    }
+
+    pub fn post_check(self) -> Result<(), AppendError> {
```

Comment:
Should we make `finish() -> Result<(), E>` a part of the visitor interface with default empty implementation. This will ensure we will not forget to call these methods manually after `accpept` and that they will be called automatically in `accept_one` too.

Reply:
That's a great idea -- Done